### PR TITLE
Fixing #9211

### PR DESCRIPTION
**Pull request by codeverifier**
- Added kubernetes version
- Fixed prettifying output
- Fixed image version display
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/9211

### CODE REVIEW

**Comment by codeverifier** [comment]
output now looks like,

```
{
  "glooVersion": {
    "client": {
      "version": "1.0.1-dev"
    },
    "server": [
      {
        "type": "Gateway",
        "enterprise": true,
        "kubernetes": {
          "containers": [
            {
              "Tag": "1.14.15",
              "Name": "observability-ee",
              "Registry": "quay.io/solo-io"
            },
            {
              "Tag": "1.14.15",
              "Name": "gloo-ee",
              "Registry": "quay.io/solo-io",
              "OssTag": "1.14.29"
            },
            {
              "Tag": "1.14.15",
              "Name": "gloo-ee-envoy-wrapper",
              "Registry": "quay.io/solo-io"
            }
          ],
          "namespace": "gloo-system"
        }
      }
    ]
  },
  "kubernetesVersion": {
    "major": "1",
    "minor": "24",
    "gitVersion": "v1.24.13+k3s1",
    "gitCommit": "3f79b289229dc0452c5cde26cccfa069805c5610",
    "gitTreeState": "clean",
    "buildDate": "2023-04-20T01:39:02Z",
    "goVersion": "go1.19.8",
    "compiler": "gc",
    "platform": "linux/amd64"
  }
}
```

File: projects/gloo/cli/pkg/cmd/version/cmd.go [hunk 1]
```diff
@@ -80,53 +99,30 @@ func getClientVersion() *version.ClientVersion {
...
 
 func printVersion(sv ServerVersion, w io.Writer, opts *options.Options) error {
-	vrs, _ := GetClientServerVersions(opts.Top.Ctx, sv)
+	vrs, k8sV, _ := GetClientServerVersions(opts.Top.Ctx, sv)
+	wrappedVersions, _ := getWrappedVersions(vrs, k8sV)
 	// ignoring error so we still print client version even if we can't get server versions (e.g., not deployed, no rbac)
 	switch opts.Top.Output {
 	case printers.JSON:
-		clientVersion, err := GetJson(vrs.GetClient())
+		formattedVer, err := getFormattedJson(wrappedVersions) // GetJson(vrs)
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/version/cmd.go:107]:
```suggestion
		formattedVer, err := getFormattedJson(wrappedVersions)
```

File: projects/gloo/cli/pkg/cmd/version/cmd.go [hunk 2]
```diff
@@ -80,53 +99,30 @@ func getClientVersion() *version.ClientVersion {
...
-			if err != nil {
-				return err
-			}
-			fmt.Fprintf(w, "%s\n", string(serverVersionStr))
+			fmt.Fprintf(w, "%s\n\n", undefinedServer)
 		}
+		fmt.Fprintf(w, "%s", string(formattedVer))
 	case printers.YAML:
-		clientVersion, err := GetYaml(vrs.GetClient())
+		formattedVer, err := getFormattedYaml(wrappedVersions) // GetJson(vrs)
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/version/cmd.go:116]:
```suggestion
		formattedVer, err := getFormattedYaml(wrappedVersions)
```

File: changelog/v1.17.0-beta11/9211-improve-glooctl-version.yaml.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+changelog:
```
[the remark this answers is not in the document]
Comment (codeverifier) [reply changelog/v1.17.0-beta11/9211-improve-glooctl-version.yaml.yaml:1]: good catch

File: projects/gloo/cli/pkg/cmd/version/clients.go [hunk 4]
```diff
@@ -110,17 +116,25 @@ func (k *kube) Get(ctx context.Context) ([]*version.ServerVersion, error) {
 			},
 		},
 	}
-	return []*version.ServerVersion{serverVersion}, nil
+	return []*version.ServerVersion{serverVersion}, k8ServerVersion, nil
```
[the remark this answers is not in the document]
Comment (codeverifier) [reply projects/gloo/cli/pkg/cmd/version/clients.go:119]: historical reasoning is beyond me. but i agree that this can be simplified.

[the remark this answers is not in the document]
Comment (jenshu) [reply projects/gloo/cli/pkg/cmd/version/clients.go:119]: according to [here](https://github.com/solo-io/gloo/blob/112ed69f2156b2456da4af6ef4f3c1fc89809adc/projects/gloo/api/grpc/version/version.proto#L55-L56) it was indeed intended to return multiple gloo instances on a cluster. it currently won't do that because if you don't pass in a namespace, it defaults to `gloo-system` and it doesn't seem like there's a way to default to "all namespaces". we _could_ fix that by allowing some way to specify "all namespaces" and return multiple gloo versions, but i think that's out of the scope of this PR

and looking at the version.proto, another possible place we could insert the k8s version is adding a new field to [Version](https://github.com/solo-io/gloo/blob/112ed69f2156b2456da4af6ef4f3c1fc89809adc/projects/gloo/api/grpc/version/version.proto#L53) - so we could have a gloo `ClientVersion`, list of gloo `ServerVersion`s, and a `KubernetesVersion`

File: projects/gloo/cli/pkg/cmd/version/clients.go [hunk 5]
```diff
@@ -41,27 +42,32 @@ func NewKube(namespace, kubeContext string) *kube {
 	}
 }
 
-func (k *kube) Get(ctx context.Context) ([]*version.ServerVersion, error) {
+func (k *kube) Get(ctx context.Context) ([]*version.ServerVersion, *kube1vVersion.Info, error) {
```
Comment (sam-heilbron) [inline projects/gloo/cli/pkg/cmd/version/clients.go:45]: Taking a look at the API that exists (https://github.com/solo-io/gloo/blob/main/projects/gloo/api/grpc/version/version.proto#L30) I think we may want to approach this slightly differently.

At the moment we have a ServerVersion concept, and there is one concrete type that we implement, a KubernetesServerVerision.

At the moment that type (https://github.com/solo-io/gloo/blob/main/projects/gloo/api/grpc/version/version.proto#L46) has a list of contains and namespace where GLoo is installed.

I think we could expand this API to also have cluster metadata. The downside is that this would lead us to duplicate the values if we did have multiple installs of gloo in the same cluster (each ServerVersion would have a reference to the same cluster metadata), but I think the benefit is that it allows us to use the existing API

**Review by sam-heilbron — COMMENTED** [review]
Thanks for tackling this! I left some thoughts about ways to expand the API using the proto as the source of truth. Let me know if that doesn't make sense and I'm happy to huddle, or outline what I had in mind

File: projects/gloo/cli/pkg/cmd/version/clients.go (hunk 5 above)
Comment (jenshu) [reply projects/gloo/cli/pkg/cmd/version/clients.go:45]: see https://github.com/solo-io/gloo/pull/9212#discussion_r1516178018, i think it might be more appropriate to put the new field (KubernetesVersion / KubernetesServerVersion) in Version, so there is just one instance returned, rather than one per gloo instance

Comment (sam-heilbron) [reply projects/gloo/cli/pkg/cmd/version/clients.go:45]: nice! yeah totally agreed with @jenshu . i didn't look far enough down in the proto 😮‍💨 , but where she proposes is exactly where I'd put it

Comment (codeverifier) [reply projects/gloo/cli/pkg/cmd/version/clients.go]: I was worried you guys might say that 😅, i just didnt want to touch the proto.
but agreed. made the change.

File: projects/gloo/cli/pkg/cmd/version/clients.go [hunk 6]
```diff
@@ -113,18 +114,75 @@ func (k *kube) Get(ctx context.Context) ([]*version.ServerVersion, error) {
...
 	splitRepoName := strings.Split(name, "/")
 	registry := strings.Join(splitRepoName[:len(splitRepoName)-1], "/")
 	img.Repository = &splitRepoName[len(splitRepoName)-1]
 	img.Registry = &registry
 	return img
 }
+
+func handleContainerImageStrFormat(str string) []string {
+	arr := strings.Split(str, ":")
+	// check for special case of image string following <registry:port/name>
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/version/clients.go:180]: it's a bit hard to follow what this function is doing. can you document here what an example input/output would be?

File: projects/gloo/cli/pkg/cmd/version/clients.go [hunk 7]
```diff
@@ -113,18 +114,75 @@ func (k *kube) Get(ctx context.Context) ([]*version.ServerVersion, error) {
...
+	if len(arr) >= 2 && strings.Index(str, "/") > strings.Index(str, ":") {
+		copyArr := make([]string, len(arr)-2)
+		copyArr[0] = arr[0] + ":" + arr[1]
```
Comment (jenshu) [inline projects/gloo/cli/pkg/cmd/version/clients.go:183]: seems to be a bug here: if `arr` is of length 2, then `copyArr` is initialized with length=0, and this line results in a panic trying to reference `copyArr[0]`

**Review by jenshu — CHANGES_REQUESTED** [review]

File: projects/gloo/cli/pkg/cmd/version/clients.go (hunk 7 above)
Comment (codeverifier) [reply projects/gloo/cli/pkg/cmd/version/clients.go:183]: good catch, il add a test case for it.

**Review by jenshu — COMMENTED** [review]
thanks for the comments and tests!

**Review by jenshu — DISMISSED** [review]

File: projects/gloo/cli/pkg/cmd/version/clients.go [hunk 8]
```diff
@@ -113,18 +114,92 @@ func (k *kube) Get(ctx context.Context) ([]*version.ServerVersion, error) {
 	return []*version.ServerVersion{serverVersion}, nil
 }
 
+func (k *kube) GetClusterVersion() (*version.KubernetesClusterVersion, error) {
+	cfg, err := kubeutils.GetConfig("", "")
+	if k.kubeContext != "" {
+		cfg, err = kubeutils.GetConfigWithContext("", "", k.kubeContext)
+	}
```
Comment (sam-heilbron) [inline projects/gloo/cli/pkg/cmd/version/clients.go:121]: These lines feel redundant. `GetConfig` just calls `GetConfigWithContext` and passes an empty string, so I think we can just simplify this to:
```suggestion
	cfg, err = kubeutils.GetConfigWithContext("", "", k.kubeContext)
```

File: projects/gloo/cli/pkg/cmd/version/version_test.go [hunk 9]
```diff
@@ -36,23 +36,47 @@ var _ = Describe("version command", func() {
 		It("will error if an error occurs while getting the version", func() {
 			fakeErr := eris.New("test")
 			client.EXPECT().Get(ctx).Return(nil, fakeErr).Times(1)
+			//client.EXPECT().GetClusterVersion().Return(nil, nil).Times(0)
```
Comment (sam-heilbron) [inline projects/gloo/cli/pkg/cmd/version/version_test.go:39]: nit: Can we delete this, or should we re-add the assertion?

**Review by sam-heilbron — DISMISSED** [review]
nice work!

File: projects/gloo/cli/pkg/cmd/version/clients.go (hunk 8 above)
Comment (codeverifier) [reply projects/gloo/cli/pkg/cmd/version/clients.go:121]: i assume this was probably left for backwards compatibility or something and never refactored. i can remove it.

**Review by jenshu — APPROVED** [review]

**Review by sam-heilbron — APPROVED** [review]

**Comment by sam-heilbron** [comment]
bulldozer
